### ISSUE TITLE
Separate typed and untyped trees

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/macros#19]
Advanced [macrology201](https://github.com/scalamacros/macrology201) covers how even simple macros can introduce owner chain corruptions that cause cryptic compiler crashes in phases after macro expansion. In [gestalt](https://github.com/liufengyun/gestalt), @liufengyun managed to automatically avoid many such problems by [separating typed and untyped trees](https://github.com/liufengyun/gestalt/pull/72) and introducing a new tree node called `TypedSplice` to embed a typed tree inside an untyped tree. The separation statically guarantees that untyped trees are not embedded in typed trees. Separation of untyped and typed trees may also offer a [solution for hygiene](https://github.com/liufengyun/gestalt/pull/77#discussion_r116548008).

**Comment by liufengyun** [comment]
The separation between typed and untyped tree originates from the design in Dotty, it helps avoids many mistakes in the compiler. Actually @odersky initially proposed the idea to macros, it turns out to be a good idea as @olafurpg mentioned (repeated below):

- auto-manage owner chains
- defend against nesting untyped trees inside typed trees
- only allow access type information on type trees
- potential to enable us do better with hygiene

As a side effect of this separation, sometimes we'll be forced to construct typed trees if we want to splice them into an existing typed tree. Constructing typed trees is like writing a compiler phase, no only there's no syntactic sugar, but also macro authors have to:

- insert type parameters manually
- insert apply manually
- insert implicits manually
- resolve overloading manually

The experiment in https://github.com/liufengyun/gestalt/pull/96 has shown it's possible to make it easier to construct typed trees by automatically infer type parameters, insert applys, insert implicits, and  resolve overloading.

**Comment by olafurpg** [comment]
Note that we may not need to expose any typed tree constructors. IIRC, your experiments in gestalt showed that splicing typed tree constructors are only necessary for few more advanced use-cases.

**Comment by liufengyun** [comment]
Yes, ideally we can have 0 extractors for untyped trees and 0 constructors for typed trees. More advanced macros would require constructors for typed trees.
